frontends/concrete-python/examples/sha256/main.py

```diff
@@ -0,0 +1,413 @@
+    if m != 0:
+        raise_low_bits = lambda x: (x % 2**m) << (CHUNK_WIDTH - m)
+        raised_low_bits = fhe.univariate(raise_low_bits)(rotated_chunks)
+        shifted_raised_low_bits = right_rotate_chunks(raised_low_bits, amount=1)
+        high_bits = rotated_chunks >> m
+        rotated_number = shifted_raised_low_bits + high_bits
+    else:
+        rotated_number = rotated_chunks
+
+    return rotated_number
+
+
+def right_shift_number(chunks, amount):
+    n = amount // CHUNK_WIDTH
+    m = amount % CHUNK_WIDTH
+
+    if n != 0:
+        shifted_chunks = right_shift_chunks(chunks, n)
+    else:
+        shifted_chunks = chunks
+
+    if m != 0:
+        raise_low_bits = lambda x: (x % 2**m) << (CHUNK_WIDTH - m)
+        raised_low_bits = fhe.univariate(raise_low_bits)(shifted_chunks)
+        shifted_raised_low_bits = right_shift_chunks(raised_low_bits, 1)
+        high_bits = shifted_chunks >> m
+        result = shifted_raised_low_bits + high_bits
+    else:
+        result = shifted_chunks
+
+    return result
+
+
+def s0(w):
+    return right_rotate_number(w, 7) ^ right_rotate_number(w, 18) ^ right_shift_number(w, 3)
+
+
+def s1(w):
+    return right_rotate_number(w, 17) ^ right_rotate_number(w, 19) ^ right_shift_number(w, 10)
+
+
+def S0(a_word):
+    return (
+        right_rotate_number(a_word, 2)
+        ^ right_rotate_number(a_word, 13)
+        ^ right_rotate_number(a_word, 22)
+    )
+
+
+def S1(e_word):
+    return (
+        right_rotate_number(e_word, 6)
+        ^ right_rotate_number(e_word, 11)
+        ^ right_rotate_number(e_word, 25)
+    )
+
+
+def Ch(e_word, f_word, g_word):
+    return (e_word & f_word) ^ ((2**CHUNK_WIDTH - 1 - e_word) & g_word)
+
+
+def Maj(a_word, b_word, c_word):
+    return (a_word & b_word) ^ (a_word & c_word) ^ (b_word & c_word)
+
+
+def loop(state, w_i_plus_k_i):
+    a, b, c, d, e, f, g, h = state
+
+    temp1 = add(h, S1(e), Ch(e, f, g), w_i_plus_k_i)
+    temp2 = add(S0(a), Maj(a, b, c))
+
+    new_a = add(temp1, temp2)
+    new_e = add(d, temp1)
+
+    return [new_a, a, b, c, new_e, e, f, g]
+
+
+def sha256(data, number_of_rounds=64):
+    h_chunks = deepcopy(H_CHUNKED)
+    k_chunks = deepcopy(K_CHUNKED)
+
+    for i in range(0, (data.shape[0] * 32) // 512):
+        with fhe.tag(f"data[{i * 16}:{(i + 1) * 16}]"):
+            start = i * 16
+            end = start + 16
+            chunks = data[start:end]
+
+            state = h_chunks
+
+            w = [None for _ in range(number_of_rounds)]
+            for j in range(0, number_of_rounds):
+                with fhe.tag(f"round-{j + 1}"):
+                    if j < 16:
+                        w[j] = chunks[j]
+                    else:
+                        w[j] = add(w[j - 16], s0(w[j - 15]), w[j - 7], s1(w[j - 2]))
+
+                    w_i_plus_k_i = add(w[j], k_chunks[j])
+                    state = loop(state, w_i_plus_k_i)
+
+            for j in range(8):
+                h_chunks[j] = add(h_chunks[j], state[j])
```

Comment:
This is very confusing overwriting both h_chunks and the state.
maybe it should be renamed round_state.
or introduce a previous_state.
```python
full_state = deepcopy(H_CHUNKED)
...
   round_state = full_state  # just an alias, full_state is not modified
   for ...:
      round_state = new_round_state(round_state, ...)
   full_state = [
      add(full_state[j], round_state[j])
      for i in range(8)
   ]
```

Reply:
It's not done in the same loop, the first loop updates `state` second loop joins `h_chunks` with `state`